chain/messagepool/repub.go

```diff
@@ -0,0 +1,106 @@
+package messagepool
+
+import (
+	"context"
+	"sort"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+const repubMsgLimit = 5
+
+func (mp *MessagePool) republishPendingMessages() error {
+	mp.curTsLk.Lock()
+	ts := mp.curTs
+	mp.curTsLk.Unlock()
+
+	baseFee, err := mp.api.ChainComputeBaseFee(context.TODO(), ts)
+	if err != nil {
+		return xerrors.Errorf("computing basefee: %w", err)
+	}
+
+	pending := make(map[address.Address]map[uint64]*types.SignedMessage)
+	mp.lk.Lock()
+	for actor := range mp.localAddrs {
+		mset, ok := mp.pending[actor]
+		if !ok {
+			continue
+		}
+		if len(mset.msgs) == 0 {
+			continue
+		}
+		// we need to copy this while holding the lock to avoid races with concurrent modification
+		pend := make(map[uint64]*types.SignedMessage, len(mset.msgs))
+		for nonce, m := range mset.msgs {
+			pend[nonce] = m
+		}
+		pending[actor] = pend
+	}
+	mp.lk.Unlock()
+
+	if len(pending) == 0 {
+		return nil
+	}
+
+	var chains []*msgChain
+	for actor, mset := range pending {
+		next := mp.createMessageChains(actor, mset, baseFee, ts)
```

Comment:
what happens if curTs changes between the top of this function and here?

Reply:
we'll use the old ts; we can hold the curTs lock longer if you think that's a problem.

Reply:
I think it makes sense to hold the lock longer. Otherwise we're gonna end up making decisions on an inconsistent state

Reply:
Sure, I'll hold it while we are also selecting the messages.
Once the messages have been selected there is no race, it is simply a matter of publishing them and it's ok if the tipset changes in the meantime.

Reply:
I extended the lock scope to include message selection.
It is does not affect `createMessageChains` as it only uses the tipset passed in.